scripts/lib/zulip_tools.py

```diff
@@ -80,6 +80,9 @@ def parse_cache_script_args(description):
     args.verbose |= args.dry_run    # Always print a detailed report in case of dry run.
     return args
 
+def get_deploy_root() -> str:
+    return os.path.realpath(os.path.dirname(os.path.dirname(os.path.dirname(__file__))))
```

Comment:
Are you sure that this is equivalent?  It really looks like it isn't; in particular, if you have a symlink you're looking at, this does't do the right thing:

```
In [2]: import os.path

In [3]: os.path.realpath("node_modules")
Out[3]: '/srv/zulip-npm-cache/74bfcc3721801df12b2065748e20e75bc4254118/node_modules'

In [4]: os.path.realpath(os.path.dirname("node_modules"))
Out[4]: '/home/tabbott/zulip'

In [5]: os.path.realpath(os.path.dirname(os.path.realpath("node_modules")))
Out[5]: '/srv/zulip-npm-cache/74bfcc3721801df12b2065748e20e75bc4254118'
```

Reply:
We’re looking at `__file__`.  `__file__` is some path ending in `scripts/lib/zulip_tools.py`. There would only be a difference if we were imported via a symlink to `scripts`, to `lib`, or to `zulip_tools.py`. Symlinks higher up in the path will be resolved equivalently either way.

Reply:
OK, I remembered what the issue was: People running scripts manually after doing `cd scripts/`.  

```
(zulip-py3-venv) tabbott@zaset:~/zulip$ cd scripts/
(zulip-py3-venv) tabbott@zaset:~/zulip/scripts$ ipython
In [1]: import os

In [2]: os.path.realpath(os.path.dirname(os.path.dirname("lib/tools.py")))
Out[2]: '/home/tabbott/zulip/scripts'

In [3]: os.path.realpath(os.path.dirname(os.path.dirname(os.path.dirname("lib/tools.py"))))
Out[3]: '/home/tabbott/zulip/scripts'

In [4]: os.path.realpath(os.path.dirname(os.path.dirname(os.path.realpath(os.path.dirname("lib/tools.py")))))
Out[4]: '/home/tabbott/zulip'
```

Reply:
(We should add this test case in a comment)

Reply:
Is `zulip_tools.py` a script that can be run?  I guess it is, though only as `zulip_tools.py make_deploy_path`, which doesn’t use `get_deploy_root`.  Is this concern worth the extra three `lstat()` calls?

Reply:
OK, I merged this commit as 70bfcd340265237add37ab15410f8dfa8dad95c3

Reply:
I guess `os.path.realpath(os.path.normpath(os.path.join(os.path.dirname(__file__), "..", "..")))` would satisfy that case without extra `lstat()`s?

Reply:
> Is zulip_tools.py a script that can be run? I guess it is, though only as zulip_tools.py make_deploy_path, which doesn’t use get_deploy_root. Is this concern worth the extra two lstat() calls?

It isn't, the actual bug is someone does `cd scripts; ./setup/install` or whatever.  This was an actual problem in production -- sysadmins who had the habit of doing `cd` before running any command would report "the script crashes".  When thousands of people have followed your installation instructions, it's definitely worth it to handle things like this correctly -- it's a big part of what makes the Zulip installation experience feel polished that almost nobody runs into issues like this in production.  

> I guess os.path.realpath(os.path.normpath(os.path.join(os.path.dirname(__file__), "..", ".."))) would satisfy that case without extra lstat()s?

Probably?  Honestly, given I'd want a little test suite (we can add one to `test-tools`) before changing this code.

Reply:
`__file__` is the path to `zulip_tools.py`, not `install`. It doesn’t matter how `install` was invoked, it only matters how `scripts.lib.zulip_tools` was imported. You noted [a problematic import](https://github.com/zulip/zulip/blob/master/scripts/zulip-puppet-apply#L8) in our conversation in personals. It’s not actually a problem in Python ≥ 3.4, though, where [`__file__` is absolute for imported modules](https://docs.python.org/3.4/whatsnew/3.4.html#other-language-changes).

Also, the outer `realpath` is redundant in any case.